File: pyfiles/psbio/stftSpectrogram.py
import warnings
import matplotlib.pyplot as plt
import numpy as np
from numpy.fft import fft, fftshift
from matplotlib import rc
from .utils import reshapeOverlap

def stftSpectrogram(x, fs, win, loverlap=0, nfft=None, zeropad=False):
    rc('font', **{'size': 18, 'family': 'serif', 'serif': ['Arial']})
    nextpow2 = lambda x: 2**np.ceil(np.log2(np.abs(x)))
    if nfft is None:
        nfft = nextpow2(len(win))

    x = reshapeOverlap(x, len(win), int(loverlap), zeropad)
    spect_vals = lambda s: np.abs(fftshift(fft(win*s, nfft)))**2
    X = np.apply_along_axis(spect_vals, 0, x) # fft along axis 0 (cols)
    X = np.apply_along_axis(lambda col: col/np.max(col), 0, X) # normalize
    X = np.apply_along_axis(lambda col: 10*np.log10(col), 0, X) # db
    X = X[0:np.floor(X.shape[0]/2).astype('int')][:] # crop negative frequencies

    fig = plt.figure()
    plt.imshow(X, cmap='jet', extent=[0, 1, 0, 1], aspect='auto', interpolation='antialiased')
    plt.xlabel('Windows [{} samples - Tres {:.3f} ms]'.format(len(win), 1000*len(win)/fs))
    plt.ylabel('Frequency (Hz/sample) - [Fres {:.3f} Hz]'.format(fs/nfft))
    plt.colorbar(label='Energy/Frequency [dB/(Hz/sample)]')
    # Ticks
    nticks = np.ceil(nfft/64).astype('int')
    ticks = np.linspace(0, 1, nticks)
    plt.xticks(ticks, np.round(ticks*X.shape[1]).astype('int'))
    plt.yticks(ticks, np.round(ticks*fs/2))
    fig.show()

    return X

File: mfiles/stftSpectrogram.m
%% stftSpectrogram: Spectrogram using Short-Time FT
function [varargout] = stftSpectrogram(x, fs, win, loverlap, nfft, zeropad)
	% Treat Inputs
	if ~exist('loverlap', 'var') || isempty(loverlap)
		loverlap = 0;
	end
	x = double(x(:));
	win = double(win(:));
	lwin = length(win);
	if ~exist('nfft', 'var') || isempty(nfft)
		nfft = lwin;
	end
	if ~exist('zeropad', 'var')
		zeropad = false;
	end

	% Reshape Signal without Zero Padding
	x = reshapeOverlap(x, lwin, loverlap, zeropad);

	% Split Apply Function
	spect_vals = @(col) 10*log10(abs(fftshift(fft(win.*col, nfft))).^2);
	x = splitapply(spect_vals, x, 1:size(x,2));
	% Remove the bottom half of the spectrogram
	x = x(1:floor(end/2),:);

	% Provide Outputs (Plot or Matrix)
	switch nargout
		case 0 % Plot Spectrogram
			figure('units','normalized','outerposition',[0 0 1 1]);
			imagesc(x);
			colormap jet;
			hcb = colorbar;
			ylabel(hcb, 'Energy/Frequency (dB/(Hz/sample))');
			set(get(gca,'YLabel'),'String', ['Frequency (Hz/sample) - Fres ' num2str(fs/nfft, 2) ' Hz']);
			ticks = linspace(1, size(x,1), 16);
			yticks('manual');
			yticks(ticks);
			fv_label = arrayfun(@num2str, round(flip(ticks)*fs/nfft, 1), 'UniformOutput', false);
			labelf = @(freq) [freq ' Hz'];
			fv_label = cellfun(labelf, fv_label, 'UniformOutput', false);
			yticklabels('manual');
			yticklabels(fv_label);
			set(get(gca,'XLabel'),'String',['Windows (' int2str(lwin) ' samples - Tres ' num2str(1000*lwin/fs, 5) ' ms)']);
			set(findall(gcf,'type','text'), 'FontSize', 22, 'fontWeight', 'bold');
			set(gca,'FontSize',22);
		case 1 % Return Spectrogram as matrix
			varargout{1} = x;
		otherwise
			error('Error in varargout! Provide the right number of outputs.');
	end
end
